### PR TITLE
feat(p2p/discovery)!: Implement `archival` discovery + syncing of historic blocks and blobs

### PR DESCRIPTION
**Pull request by renaynay**
This PR introduces the discovery of nodes advertising on the `archival` topic for the purpose of being able to request blobs and blocks that are older than the sampling window. 

In order to do so, there are now 2 instances of discovery (one for `full` nodes -- _all nodes running full block sync within the sampling window_, and one for `archival` nodes -- _nodes that sync, store and serve all historic data_) and 2 instances of peer managers (one for `full` and one for `archival`).

**This PR is _breaking_ due to changes to metrics names.**

Please also note that this PR introduces the ability to perform archival sync of blobs + blocks for **LIGHT AND FULL** nodes only regardless of whether the full nodes have pruning enabled or not. **Bridge nodes that have previously had pruning enabled will NOT be able to request historic data from the consensus core node.** The bridge node must run in archival mode (without `--experimental-pruning` enabled) in order to access historic data.

**Bridge and Full nodes in archival mode (without `--experimental-pruning`)**
```
2024-05-10T15:21:44.490+0200	INFO	share/discovery	discovery/manager.go:30	starting discovery	{"topic": "full"}
2024-05-10T15:21:44.491+0200	INFO	share/discovery	discovery/manager.go:33	advertising to topic	{"topic": "full"}
2024-05-10T15:21:44.491+0200	INFO	share/discovery	discovery/manager.go:30	starting discovery	{"topic": "archival"}
2024-05-10T15:21:44.491+0200	INFO	share/discovery	discovery/manager.go:33	advertising to topic	{"topic": "archival"}
```

**Bridge and Full nodes running `--experimental-pruning`**
```
2024-05-10T15:19:20.698+0200	INFO	share/discovery	discovery/manager.go:30	starting discovery	{"topic": "full"}
2024-05-10T15:19:20.698+0200	INFO	share/discovery	discovery/manager.go:33	advertising to topic	{"topic": "full"}
2024-05-10T15:19:20.698+0200	INFO	share/discovery	discovery/manager.go:30	starting discovery	{"topic": "archival"}
```

**Light nodes**
```
2024-05-10T15:20:42.440+0200	INFO	share/discovery	discovery/manager.go:30	starting discovery	{"topic": "full"}
2024-05-10T15:20:42.440+0200	INFO	share/discovery	discovery/manager.go:30	starting discovery	{"topic": "archival"}
```

-----------------------------------------------------------

**TODO**
- [x] test archival request routing per node type
- [x] renaming some things inside discovery away from being `full`-node-centric
- [x] rebase
- [x] swamp test

For follow-up PR:
- [x] clarify / make disc metrics generic as there will now be 2 instances of discovery happening

Nice to have; 
- [ ] benchmarks for LN

### CODE REVIEW

File: nodebuilder/share/module.go
Comment (distractedm1nd) [inline nodebuilder/share/module.go:306]: Don't pruned nodes at least need archivalDisc?

Comment (renaynay) [reply nodebuilder/share/module.go]: here it is: https://github.com/celestiaorg/celestia-node/pull/3188/commits/9604b96b7c9aa6fcf7fbfca31db11f4906487584

File: nodebuilder/share/module.go
Comment (Wondertan) [inline nodebuilder/share/module.go:362]: The fact that we need to hold the map in FX already suggests adding a higher-level component. Something like DiscoveryManager that has discovery CRUD (or at least CR portion for the start.) This will decrease the complexity of node builder setup and avoid hassle with FX

File: nodebuilder/node.go
Comment (renaynay) [inline nodebuilder/node.go:61]: This is needed for swamp testing (ability to look into EDSStore the same way we can do with header store).

File: nodebuilder/share/module.go
Comment (renaynay) [reply nodebuilder/share/module.go:359]: will consider it

File: share/p2p/discovery/manager.go
Comment (renaynay) [inline share/p2p/discovery/manager.go:1]: I didn't know where else to leave this comment so I'll do it here: 

Please note that there is one discovery parameters that is shared between both archival + full discovery instances. Right now I don't want to manage several configs per discovery type so i'm just going to leave it as is, but it's just something to note that both instance of discovery are tied to the same configuration options (minus `EnableAdvertise` bc I configure that per instance in nodebuilder).

Comment (renaynay) [reply share/p2p/discovery/manager.go:1]: Also the separation between Parameters + options in discovery is kind of ugly. I propose to clean this up in later PRs as tech debt otherwise this PR will continue to grow in scope. 

(e.g. I'd like to clean up the way onPeersUpdate works so that I don't have to do annoying magic with fx to get it set up for archival peer manager)

**Comment by walldiss** [comment]
Great work on the PR. I had a thought after going through it—what if we tried a slightly different approach with the routing and construction? We could introduce something like a PrunedGetter that utilizes two getters under the hood: pruned and archival. It would automatically route requests based on availability, which could help keep the routing logic out of shrex.getter, especially since it’s getting a bit complex. This setup might also streamline construction—you’d just need to set up the archival and pruned shrex getters independently and then link them to the PrunedGetter. Plus, each getter would manage its own peers and discovery, simplifying the need for a complex discovery manager. I think this could make future changes and maintenance a lot easier. Just a thought to consider! 😊

**Comment by renaynay** [comment]
@walldiss This is a good suggestion and I share your concern for dirtying the shrex getter but I think that future-proofing it for other kinds of peer routing is out of scope for this PR (especially given how long it's taking to get it out the door).

I've made an issue describing the problem here and i think it could be worked on as part of partial nodes workstream once that is approved / kicked off: 

https://github.com/celestiaorg/celestia-node/issues/3393

File: share/p2p/discovery/manager.go
Comment (distractedm1nd) [reply share/p2p/discovery/manager.go:1]: I agree with Vlad on this point. I would be comfortable shipping like this given that you or someone else already commits to working on it next

File: nodebuilder/pruner/module.go
Comment (renaynay) [reply nodebuilder/pruner/module.go:88]: Where are you referring?

File: share/p2p/discovery/manager.go
Comment (renaynay) [reply share/p2p/discovery/manager.go:27]: Changed it a bit and also cancel the ctx if there's an error to short-circuit out of advertise loop in case of err

File: nodebuilder/tests/prune_test.go
Comment (walldiss) [inline nodebuilder/tests/prune_test.go:31]: Based on the description, it looks like those cases are covered:
- Sync from archival_BN to archival_FN
- Sync from archival_BN+pruning_BN to pruning_FN
- LN sampling from archival_FN

What is the purpose of the second test? It doesn’t seem to be testing anything meaningful. Perhaps it should be broken into separate test cases to isolate routing conditions:

- Pruning_FN can sync from only pruning_FN/BN.
- Pruning_FN from only archival_FN.

Also missing:
- Archival_FN from archival_FN. There is a test syncing from archival_BN, which is similar but not the same.

In the current setup, LN requests from both archival and pruning run at the same time. I would also suggest breaking the test into two:
- LN can manually request archival and recent blobs from ONLY archival_FN.
- LN can manually request recent blobs from ONLY pruning_FN.

File: share/p2p/discovery/manager.go
Comment (walldiss) [inline share/p2p/discovery/manager.go:27]: If there are multiple discoveries and last one failed the function will return. Perhaps started ones should be stopped before exit.

File: share/p2p/discovery/manager.go
Comment (walldiss) [inline share/p2p/discovery/manager.go:25]: Please, use context from Start function call instead of the Advertisement context(WithCancel).

File: share/p2p/discovery/metrics.go
Comment (walldiss) [inline share/p2p/discovery/metrics.go:56]: Nit: could be just:
```
d.tag + "_discovery_amount_of_peers"
```
I personally find it cognitively harder to parse template.

**Review by walldiss — COMMENTED** [review]
Can't tell if nodebuilder is correct. Changes are quite complex and difficult to fully understand. So relying on tests for correctness. Not a good thing, so we should prioritise refactoring of share module construction.

File: nodebuilder/pruner/module.go
Comment (vgonkivs) [reply nodebuilder/pruner/module.go:88]: https://github.com/celestiaorg/celestia-node/pull/3188/files#diff-0b8ebf8b9930a092b6ce9bc7d7c9a95c638c5772d452ba2241a5d344c610fdabR70-R72

File: nodebuilder/pruner/module.go
Comment (renaynay) [reply nodebuilder/pruner/module.go:105]: Yes unfortunately for the bridge node's core components.

File: share/p2p/discovery/manager.go
Comment (renaynay) [reply share/p2p/discovery/manager.go:27]: this could be resolved if fx called Stop if Start partially fails on the node

File: share/p2p/discovery/manager.go
Comment (renaynay) [reply share/p2p/discovery/manager.go:25]: context from start function is short lived

Comment (walldiss) [reply share/p2p/discovery/manager.go:25]: yes, and the Start context is meant to be passed to all Start calls

File: nodebuilder/share/module.go
Comment (renaynay) [reply nodebuilder/share/module.go]: yea also realised

File: nodebuilder/share/constructors.go
Comment (renaynay) [reply nodebuilder/share/constructors.go]: it's `FullDiscovery` bc it's discovering fulls

File: nodebuilder/pruner/module.go
Comment (renaynay) [reply nodebuilder/pruner/module.go]: tried to clean it up

File: nodebuilder/tests/prune_test.go
Comment (renaynay) [reply nodebuilder/tests/prune_test.go:31]: > What is the purpose of the second test? It doesn’t seem to be testing anything meaningful. 

which second test do you mean? There is only one test here that tests to ensure a LN can discover + sync historic blobs off of an archival node in a network topology where pruned FNs are dominant.

> LN can manually request recent blobs from ONLY pruning_FN

We can't control that the LN requests recent blobs only from pruned FNs as archival nodes can serve the exact same data and also advertise as fulls in the network.

Comment (renaynay) [reply nodebuilder/tests/prune_test.go:31]: I agree we should add an additional test case for a pruned FN requesting historic data from archival FN/BN though, but I'd like to put that in an issue and address it after this PR.

File: share/p2p/discovery/manager.go
Comment (renaynay) [reply share/p2p/discovery/manager.go:25]: Okay I will, but that context is ignored inside discovery.Start anyway in favour of an internal context. It just spawns a bunch of routines.

File: share/p2p/discovery/manager.go
Comment (renaynay) [reply share/p2p/discovery/manager.go:27]: https://github.com/celestiaorg/celestia-node/issues/3408

File: share/p2p/discovery/manager.go
Comment (Wondertan) [reply share/p2p/discovery/manager.go:1]: Which point tho?

File: share/p2p/discovery/manager.go
Comment (Wondertan) [reply share/p2p/discovery/manager.go:8]: I really don't want to block on that as this PR is long overdue, but the current state is pretty messy and feels even messier since the last look I made. One of the options I see is to roll back to the last pre-refactoring version, and park cleans up in the issue. If we you see we can have some more time, then better doing it now ofc.

**Review by Wondertan — COMMENTED** [review]
There are two main comments. The introduced Manager is net negative and the need for EnableAdvertise in config is unclear.

File: share/p2p/discovery/manager.go
Comment (Wondertan) [reply share/p2p/discovery/manager.go:27]: Likely unneeded with approach described in https://github.com/celestiaorg/celestia-node/pull/3188#discussion_r1604349569

File: share/getters/shrex.go
Comment (walldiss) [inline share/getters/shrex.go:302]: Related to optional archivalPeerMan: Lets remove nil check.
```
sg.archivalPeerManager != nil 
```

File: share/getters/options.go
Comment (walldiss) [inline share/getters/options.go:20]: Looking at this codes makes me wonder what is the case when archivalPeerManager is nil? It seems it should always be provided, to allow shrex.Getter to access historcal data. Lets make it archivalPeerManager and availabilityWindow part of constructor and not an option

File: share/getters/shrex.go
Comment (walldiss) [inline share/getters/shrex.go:302]: Also if sg.availabilityWindow == 0, it will always result in call of `pruner.IsWithinAvailabilityWindow` to return true, even for historical. This default works well for DASer, but is incorrect here. This default setting will result in all request to be routed to prunned nodes. Perhaps lets disable sg.availabilityWindow == 0 being possible on params validation at least, to ensure misconfiguration does not happen.

File: nodebuilder/tests/prune_test.go
Comment (walldiss) [reply nodebuilder/tests/prune_test.go]: >which second test do you mean? There is only one test here that tests to ensure a LN can discover + sync historic blobs off of an archival node in a network topology where pruned FNs are dominant.

Sorry for misleading wording, I meant the second case I've listened in covered cases. I. referred to  it as second test, because single test covers multiples scenarios.

>We can't control that the LN requests recent blobs only from pruned FNs as archival nodes can serve the exact same data and also advertise as fulls in the network.

Imagine pruned are not available and it always falls back to archival serving. We need the test to ensure pruned can serve too. You can shutdown archival to ensure pruned are able to serve.

File: share/p2p/discovery/manager.go
Comment (walldiss) [reply share/p2p/discovery/manager.go:25]: discovery.Start call does not use context indeed, but it expects short lived one. What is the reason to use incorrect context here?

File: nodebuilder/share/opts.go
Comment (cristaloleg) [inline nodebuilder/share/opts.go:19]: Nit: looks like this has `O(N**2)` complexity (see https://cs.opensource.google/go/go/+/refs/tags/go1.22.3:src/errors/join.go;l=19)

Not sure how important it's but collecting errors and than joining them looks simpler. Same in the `func` below and in `share/p2p/discovery/manager.go`

File: share/getters/options.go
Comment (renaynay) [reply share/getters/options.go:20]: At the time, I was unsure about whether every node needs to have an archival service, but now we said that they need to.

File: share/getters/shrex.go
Comment (renaynay) [reply share/getters/shrex.go]: @walldiss full peers are not necessarily pruned 

archival peers also advertise on `full` topic.

File: nodebuilder/share/module.go
Comment (renaynay) [reply nodebuilder/share/module.go]: Reworked in https://github.com/celestiaorg/celestia-node/pull/3188/commits/aa091af3741143fba0305ef5d4fecd7d4f566a28

File: share/p2p/discovery/options.go
Comment (renaynay) [reply share/p2p/discovery/options.go]: https://github.com/celestiaorg/celestia-node/pull/3188/commits/aa091af3741143fba0305ef5d4fecd7d4f566a28

File: nodebuilder/share/module.go
Comment (walldiss) [inline nodebuilder/share/module.go:148]: shrexSub subscriber should be part of Full and Light and not be part of Bridge. Somehow it is the opposite in this constructor

File: nodebuilder/share/module.go
Comment (walldiss) [inline nodebuilder/share/module.go:233]: Bridge does not need availability

File: nodebuilder/share/module.go
Comment (walldiss) [inline nodebuilder/share/module.go:237]: Consider returning error like "Node type: %s is not supported for availability"

File: nodebuilder/share/opts.go
Comment (walldiss) [inline nodebuilder/share/opts.go:16]: What is the reason to use map[string]*peers.Manager? Is it just the way to pass multiple components of same type?

File: share/getters/options.go
Comment (walldiss) [inline share/getters/options.go:13]: I think it should be not an options, but part of constructor

File: share/p2p/discovery/manager.go
Comment (walldiss) [inline share/p2p/discovery/manager.go:27]: Lets make an issue about stopping started discoveries if error happen

File: share/p2p/peers/options.go
Comment (walldiss) [inline share/p2p/peers/options.go:91]: Should be part of the constructor, because it is not optional - needs to be set explicitly.

File: share/getters/shrex.go
Comment (renaynay) [reply share/getters/shrex.go]: I still don't understand why the logic here isn't applicable. `full` topic contains *all* FN + BN in the network, not just pruned, so why would routing peer requests to `full` peers be an issue? It's how the network worked before as well. If it's just semantics, then maybe we can make an issue for it?

File: nodebuilder/share/module.go
Comment (renaynay) [reply nodebuilder/share/module.go:148]: @walldiss shrexsub subscriber lifecycle is unfortunately implicitly managed inside peer manager, but we do not pass shrexsub to peer manager for bridge nodes bc we do not care to listen to shrexsub pools. BN still needs to broadcast to shrexsub though so we need to manage shrexsub lifecycle here.

File: nodebuilder/share/module.go
Comment (renaynay) [reply nodebuilder/share/module.go:233]: It needs it for the top level share module.

File: nodebuilder/share/module.go
Comment (renaynay) [reply nodebuilder/share/module.go:237]: we do this everywhere else.

File: nodebuilder/share/opts.go
Comment (renaynay) [reply nodebuilder/share/opts.go:16]: we provide this as a type to fx bc we need to pass in the managers in the correct order to shrex getter, so unless I provide them as an []*peers.Manager as well, this will have to do :/

File: share/p2p/discovery/manager.go
Comment (renaynay) [reply share/p2p/discovery/manager.go:27]: We should make it for all services but yes.

File: share/getters/options.go
Comment (renaynay) [reply share/getters/options.go]: https://github.com/celestiaorg/celestia-node/pull/3188/commits/da20efc103bf02797140b067d0913396fc196089

File: nodebuilder/share/module.go
Comment (renaynay) [reply nodebuilder/share/module.go:151]: https://github.com/celestiaorg/celestia-node/pull/3188/commits/da20efc103bf02797140b067d0913396fc196089

this unlocked new change though

File: share/p2p/discovery/manager.go
Comment (renaynay) [reply share/p2p/discovery/manager.go:27]: #3408

File: share/p2p/peers/options.go
Comment (renaynay) [reply share/p2p/peers/options.go]: https://github.com/celestiaorg/celestia-node/pull/3188/commits/b48d69673b41689be9a2ff4012c9e4a20c35649c

File: nodebuilder/share/module.go
Comment (walldiss) [reply nodebuilder/share/module.go:151]: Ughhh. Good to know. Who might have though it is good idea to manage lifecycle in peer-manager 🥇 https://github.com/celestiaorg/celestia-node/issues/3433
I'm fine if it will be fixed later.

File: nodebuilder/share/module.go
Comment (walldiss) [reply nodebuilder/share/module.go:236]: Oh, yeah, it is used for Availability rpc.

File: share/p2p/discovery/manager.go
Comment (walldiss) [reply share/p2p/discovery/manager.go:27]: Well, good to know that we don't stop node on fail. Problem with partially started discoveries will still persist, even if mentioned issue is implemented. So needs to land in mentioned/new issue or to be resolved with modified Stop here.

File: nodebuilder/share/opts.go
Comment (walldiss) [reply nodebuilder/share/opts.go:16]: Thank you! It didn't clicked for me initially, but now I got the idea!

File: share/getters/shrex.go
Comment (walldiss) [reply share/getters/shrex.go]: Because sg.availabilityWindow == 0 disables routing to prunned nodes. You never want to disable routing here, regardless of local availability window setup. 
Perhaps confusion comes from the fact that it is configurable in the first place, when it should be just a constant.

File: share/p2p/discovery/discovery.go
Comment (Wondertan) [inline share/p2p/discovery/discovery.go:55]: This is not used anyhow, also comments needs a fix

Comment (Wondertan) [reply share/p2p/discovery/discovery.go:55]: I guess the idea was to avoid advertising all discoveries at once, but left unfinished

Comment (renaynay) [reply share/p2p/discovery/discovery.go:55]: It is used, good catch on the comment. I added advertise to the private options (where the callback func for peer manager exists as well)

**Review by Wondertan — APPROVED** [review]
I want to ship pruning and unblock this PR, so approving, 
Under different circumstances, I would request a redesign.
This PR leaves too much dech debt behind it

File: share/p2p/discovery/discovery.go
Comment (Wondertan) [reply share/p2p/discovery/discovery.go]: I see, it was used in the manager accessing the private field :sigh:

**Review by walldiss — APPROVED** [review]
Great effort on the nodebuilder in this PR! However, it's still pretty tough to tell if everything's correct. My biggest concern is that the feature needs thorough testing to ensure all setups work as intended. Given its complexity, future changes might get tricky, so let's prioritize refactoring as soon as we have the resources.

**Review by Wondertan — APPROVED** [review]

**Pull request merged**
